### ISSUE TITLE
Add label to provider components for easier clusterctl adoption

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/cluster-api#1989]
**User Story**

As an operator I would like to start using clusterctl with my existing management cluster because clusterctl is awesome  😉 

**Detailed Description**

When installing a provider, clusterctl v2 makes some opinionated choices to simplify the management of upgrades/other lifecycle operations.

One of these choices is to mark all the provider components with the `clusterctl.cluster.x-k8s.io/provider` label.

If we can start adding a similar label to the next v1alpha2 patch versions of existing providers, this will make it easier to start using the new clusterctl tool on the existing management clusters. 

More specifically this will enable usage of clusterctl for v1alpha2 --> v1alpha3 upgrade.

e.g. 
for CAPI: `clusterctl.cluster.x-k8s.io/provider=aws`
for CABPK: `clusterctl.cluster.x-k8s.io/provider=kubeadm-bootstrap`
for CACPK: `clusterctl.cluster.x-k8s.io/provider=kubeadm-controlplane`
for CAPA: `clusterctl.cluster.x-k8s.io/provider=aws`
for CAPV: `clusterctl.cluster.x-k8s.io/provider=vsphere`
for CAPD: `clusterctl.cluster.x-k8s.io/provider=docker`

/kind feature
/area clusterctl 

/cc @vincepri @detiber @akutz @chuckha @ncdc

**Comment by wfernandes** [comment]
@fabriziopandini I would like to work on this issue. Unfortunately, I don't have much context on what needs to be done here so any direction would be much appreciated. 
I also noticed in the description we have: 
> for CAPI: clusterctl.cluster.x-k8s.io/provider=aws

Did you want for CAPI to be `core` or `capi`?

**Comment by fabriziopandini** [comment]
@wfernandes thanks for helping!
The goal of this change is to add the two labels to all the resources in the manifest generated for the v1alpha2  branch and for master.

I'm not a kustomize expert, but AFAIK it is not possible to use the `commonLabels` option in the top-level kustomization file under `config/default`, because this will apply the labels also to label selectors, which is something I'm pretty sure we don't want.  
However, I think it is possible to use `commonLabels` in all the imported kustomization files (e.g. `config/rbac/kustomization.yaml`) with the exception on `config/manager` where you have to apply labels object per object in order to avoid to get the label selector changed.

> Did you want for CAPI to be core or capi?

cluster-api

**Comment by wfernandes** [comment]
@fabriziopandini I noticed in the code that we are adding the labels `clusterctl.cluster.x-k8s.io=` and `clusterctl.cluster.x-k8s.io/provider=cluster-api` when creating/updating the component objects. 
So if we keep this logic it will override whatever labels would have been generated as part of the kustomize manifests. 

https://github.com/kubernetes-sigs/cluster-api/blob/0f62f835a81f173dffba4b5fc24f54a0e8d7c5cd/cmd/clusterctl/pkg/client/repository/components.go#L521-L542

Is this behavior intentional?
Are we planning on deprecating the `addLabels` logic?

**Comment by fabriziopandini** [comment]
Adding labels to the manifest is meant to simplify starting using clusterctl on a cluster created with kubectl apply manifest (without clusterctl).

The fact that when you install the same manifest clusterctl overrides the existing labels is acceptable from my point of view, but I'm not dropping the addLabels logic because there is no guarantee that all the providers are going to add labels to the generated manifests

**Comment by wfernandes** [comment]
Todos:
- [x] Add label to CAPI
- [x] Add label to CABPK
- [x] Add label to CACPK
- [ ] Add label to CAPD
- [ ] Add label to CAPA
- [ ] Add label to CAPV
- [ ] Backport all the changes to v1alpha2 so that v1apha2 --> v1alpha3 upgrade is supported.

**Comment by wfernandes** [comment]
@fabriziopandini Regarding the following
> If we can start adding a similar label to the next v1alpha2 patch versions of existing providers, this will make it easier to start using the new clusterctl tool on the existing management clusters.

Do I need to backport the changes to `release-0.2` or do I need to add some conversion logic for v1alpha2?

**Comment by fabriziopandini** [comment]
@wfernandes yes, thanks

**Comment by fabriziopandini** [comment]
@wfernandes as per zoom discussion:

- user facing documentation should go [here](https://master.cluster-api.sigs.k8s.io/clusterctl/commands/adopt.html): 
TL;DR; Labels are a precondition for the adopt command; Users can get labels for free installing latest v1alpha2/v1alpha3 manifests or add labels manually.
- documentation for providers author should go in a new "common labels" paragraph under the [Components YAML](https://master.cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#components-yaml) section: 
TL;DR; the component YAML should always add a label so it will be easier to make the transition from kubectl apply to clusterctl
- documentation about common providers name should go in the [Provider repositories](https://master.cluster-api.sigs.k8s.io/clusterctl/configuration.html#provider-repositories) section after "By default, clusterctl ships with providers sponsored by SIG Cluster Lifecycle."
